config/src/main/java/org/springframework/security/config/http/DefaultFilterChainValidator.java

```diff
@@ -88,10 +88,13 @@ private void checkForDuplicateMatchers(List<SecurityFilterChain> chains) {
 		while (chains.size() > 1) {
 			DefaultSecurityFilterChain chain = (DefaultSecurityFilterChain) chains.remove(0);
 			for (SecurityFilterChain test : chains) {
-				if (chain.getRequestMatcher().equals(((DefaultSecurityFilterChain) test).getRequestMatcher())) {
-					throw new IllegalArgumentException("The FilterChainProxy contains two filter chains using the"
-							+ " matcher " + chain.getRequestMatcher() + ". If you are using multiple <http> namespace "
-							+ "elements, you must use a 'pattern' attribute to define the request patterns to which they apply.");
+				if (test instanceof DefaultSecurityFilterChain securityFilterChain) {
+					if (chain.getRequestMatcher().equals(securityFilterChain.getRequestMatcher())) {
+						throw new IllegalArgumentException("The FilterChainProxy contains two filter chains using the"
```

Comment:
These message is XML-specific, so they arent't as relevant to Java Config users. Still, I think that each application type should see error messages that make sense to them.

To address this, perhaps this class can contain a private inner interface with methods like:

```java
void handleDuplicateRequestMatchers(chain, securityFilterChain);

void handleChainAfterAnyRequest(anyRequestChain, afterChain);
```

(don't worry so much about the names, just giving the idea)

Then, internally there would be a XML and Java Config versions of the interface. This would allow each message to be relevant to the application type. There may be other ways to achieve this as well. What I'd prefer to avoid is adding `if/else` statements throughout error handling logic.

Reply:
For example, strategy:
```
interface RequestMatchersCheckingStrategy {
		void handleDuplicateRequestMatchers(SecurityFilterChain chain, SecurityFilterChain securityFilterChain);

		void handleChainAfterAnyRequest(SecurityFilterChain anyRequestChain, SecurityFilterChain afterChain);
	}
```
And its two implementations `XmlRequestMatchersCheckingStrategy` and `JavaRequestMatchersCheckingStrategy`. The `checkForDuplicateMatchers` method will look like:
```
private void checkForDuplicateMatchers(List<SecurityFilterChain> chains) {
		while (chains.size() > 1) {
			DefaultSecurityFilterChain chain = (DefaultSecurityFilterChain) chains.remove(0);
			for (SecurityFilterChain test : chains) {
				requestMatchersCheckingStrategy.handleDuplicateRequestMatchers(test, chain);
			}
		}
	}
```
I’ll just note that this interface must be public in order to be able to specify its implementation:
```
filterChainValidator.setRequestMatchersCheckingStrategy(new JavaRequestMatchersCheckingStrategy());
```